### ISSUE TITLE
Add option to teardown and re-setup tests between runs

### ISSUE THREAD

**Issue by mikeybinns** [vitest-dev/vitest#3716]
### Clear and concise description of the problem

There's currently no way to run a function when vitest triggers a re-test based on changed watched files. It used to be possible to hack this in with reporters, but with the introduction of workspaces, this doesn't work for workspaces.

### Suggested solution

Some possible options:

1. Add extra options to watch config:
```
// as well as
watch: true,
// also allow
watch: {
  rerunTests: true, // the same as the boolean above
  onWatcherRerun: async () => {}, // The function to run between reruns
  rerunSetupOnChange: true, // Shortcut to automatically teardown and setup as per already configured parameters.
}
```
2. Add onWatcherRerun to projects as per the above setup, but as a root config/project configuration parameter.

### Alternative

For my specific use case, I'm testing an npm package, so I need to use setup to build the package and add it to a test project, so I can use it the same way I would use any package normally, but every change I make must also rebuild the package again.

For reference, if you aren't using workspaces, you can do this:
```
import { defineConfig } from "vitest/config";
import { setup, teardown } from "./src/test-setup";

export default defineConfig({
  test: {
    include: ["src/**/*.test.{js,mjs,cjs,ts,jsx,tsx}"],
    watchExclude: ["**/node_modules/**", "**/dist/**"],
    globalSetup: ["./src/test-setup.ts"],
    reporters: [
      "default",
      {
        async onWatcherRerun() {
          // Delete test package and node_modules
          await teardown();
          // re-create new test package based on new build
          await setup();
        },
      },
    ],
  },
});
```

### Additional context

I found issue #2614 but this seems to be about triggering reruns when setup files change.

I realise that for most use cases, this would be opposed to what you want, because it would make tests slower, so this shouldn't be a default, but having an option to allow it would be great for edge use cases where setup must be done before a test can be done.

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

**Comment by sheremet-va** [comment]
What do you mean it doesn't work after workspaces were introduced? What's different now?

**Comment by mikeybinns** [comment]
Sorry, that's actually a miscommunication on my part. With workspaces, you can only specify reporters on the whole monorepo level, you can't specify them for individual workspaces/projects. This will still work at the global level, but it would slow down all my packages tests, not just the single package that actually requires it.

So say I have a workspace with packages A, B and C.
If package A requires the re-setup, but B and C don't, it would still rerun the setup if I changed a file within package B or C, which seems opposed to how the workspaces feature should work.

**Comment by sheremet-va** [comment]
> Sorry, that's actually a miscommunication on my part. With workspaces, you can only specify reporters on the whole monorepo level, you can't specify them for individual workspaces/projects. This will still work at the global level, but it would slow down all my packages tests, not just the single package that actually requires it.
> 
> So say I have a workspace with packages A, B and C. If package A requires the re-setup, but B and C don't, it would still rerun the setup if I changed a file within package B or C, which seems opposed to how the workspaces feature should work.

Watch mode works on a workspace level, so this is why you can only access it in workspace reporters. We can’t really implement it other way, but you can filter it yourself in the reporter I think. onWatchRerun receives a lit of files I think

**Comment by sheremet-va** [comment]
The fact that global setup/teardown is running is a bug and there is a separate issue for this, if you have s problem with that specific usecase.

**Comment by mikeybinns** [comment]
Sorry, I think I'm confusing things with my terminology. For the following, assume `./` refers to the monorepo root.

In an indivudual package within a monorepo, you would call `defineProject` (`./packages/a/vitest.config.ts`), this is where I want to specify my `onWatchRerun` function, but I can't specify it because `defineProject` doesn't allow you to pass `reporters`. I also cannot specify `reporters` in a project instance in `defineWorkspace` (`./vitest.workspace.ts`) for the same reason. 
```
Object literal may only specify known properties, and 'reporters' does not exist in type 'ProjectConfig'.
```

I can specify `reporters` in the monorepo config (`./vitest.config.ts`), which is where I would call `defineConfig` but this would apply to all tests in all packages in my monorepo, so if I changed code in package B, it would run the `onWatchRerun` even though I don't want it to.

One thing I just thought of as a workaround is maybe I could define it in the monorepo config and just do a check to see if the edited files are part of a certain package, but this is not ideal because then I'm specifying package level configuration outside of the package config, and I'd prefer to keep all of that within the project config.

**Comment by mikeybinns** [comment]
Sorry. I misread what you said and then posted a comment that just basically re-worded what you said.

So watching works at a workspace level meaning if I change a file in package B, it would rerun tests in package A as well?

**Comment by sheremet-va** [comment]
> Sorry. I misread what you said and then posted a comment that just basically re-worded what you said, so I deleted it.
> 
> So watching works at a workspace level meaning if I change a file in package B, it would rerun tests in package A as well?

It’s possible to have the same test in different workspace projects, so yes we still check the config when file changes. It might not rerun tests themselves if they were not changed.

But watching works only on root level and its not possible to have watch behavior configured per project.

**Comment by mikeybinns** [comment]
Okay, no problem, thanks :)
